pulsar/internal/backoff.go

```diff
@@ -18,28 +18,64 @@
 package internal
 
 import (
+	"math/rand"
 	"time"
 )
 
-// Backoff
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
 type Backoff struct {
-	backoff time.Duration
+	initGap, nextGap, maxGap time.Duration
+
+	firstRetry       time.Time
+	mandatoryStop    time.Duration
+	mandatoryStopped bool
 }
 
-const (
-	minBackoff = 100 * time.Millisecond
-	maxBackoff = 60 * time.Second
-)
+func NewBackoff(initGap, maxGap, mandatoryStop time.Duration) *Backoff {
+	return &Backoff{
+		initGap:          initGap,
+		nextGap:          initGap,
+		maxGap:           maxGap,
+		firstRetry:       time.Time{},
+		mandatoryStop:    mandatoryStop,
+		mandatoryStopped: false,
+	}
+}
+
+// for test mock convenience
+type nowTimeProvider func() time.Time
+
+var systemNow nowTimeProvider = func() time.Time {
+	return time.Now()
+}
 
-// Next
 func (b *Backoff) Next() time.Duration {
-	// Double the delay each time
-	b.backoff += b.backoff
-	if b.backoff.Nanoseconds() < minBackoff.Nanoseconds() {
-		b.backoff = minBackoff
-	} else if b.backoff.Nanoseconds() > maxBackoff.Nanoseconds() {
-		b.backoff = maxBackoff
+	curGap := b.nextGap
+	if curGap < b.maxGap {
+		b.nextGap = MinDuration(2*b.nextGap, b.maxGap)
 	}
 
-	return b.backoff
+	if b.mandatoryStop > 0 && !b.mandatoryStopped {
+		now := systemNow()
+		tried := time.Duration(0)
+		if curGap == b.initGap {
+			b.firstRetry = now
+		} else {
+			tried = now.Sub(b.firstRetry)
+		}
+		if tried+curGap > b.mandatoryStop {
+			// reached mandatory stop, reset current retry gap
```

Comment:
If mandatoryStop set to 0, this always be true in first time, means there is no mandatory stop which is on expected.
So I rule it out in above `if b.mandatoryStop > 0`, more readable.